lib/src/environment/controls/environment_control.dart

```diff
@@ -0,0 +1,71 @@
+import 'package:flutter/widgets.dart';
+
+import '../state/environment_state.dart';
+
+/// Signature for a function that creates an EnvironmentState-backed widget.
+typedef EnvironmentControlBuilder<T> = Widget Function(
+  BuildContext context,
+  T? currentValue,
+  Function(T) updateValue,
+);
+
+/// Provides a widget that allows interaction with an [EnvironmentState]-backed
+/// variable.
+abstract class EnvironmentControl<T> {
```

Comment:
I like this overall! My only vague feedback is around whether it's possible to push those closer to a pure interface, just so that it is more difficult to screw up overriding the wrong methods, forgetting to call super, all the usual "composition over inheritance" arguments. The more I think about this though, if we made this a pure interface, it seems like we'd still want a base class or helper class with convenience logic for updateValue, currentValue etc, so this seems good and practical.

Reply:
My best crack at a composition-based solution is for EnvironmentControl to accept something like a KeyedEnvironmentState<T> which essentially contains the logic in updateValue and currentValue. E.g.

```
abstract class KeyedEnvironmentState<T> {
  T get value;

  // The framework calls onValueUpdated on your control when this is called.
  void updateValue(T value)
}

class BooleanControl {
  BooleanControl({
    required KeyedEnvironmentState<boolean> state,
    required String title,
    ...
  })
}

final boldControl = BooleanControl(
   state: environmentState.forKey<boolean>('Controls.bold'),
   title: 'Bold control',
   ...
)
```


Overall, the above seems maybe more complex and confusing than what you have now, so let's stick with what you have!

Reply:
Yeah, there's a good amount of shared functionality that goes into the base class that I think would be really easy to forget in implementations (source: I did this several times before moving the logic into a base class).